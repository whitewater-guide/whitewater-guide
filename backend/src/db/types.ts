export interface RawTimestamped {
  created_at: Date;
  updated_at: Date;
}

/**
 * This is result of graphql query
 * It will convert Dates to strings
 */
export interface Timestamped {
  createdAt: string;
  updatedAt: string;
}

export const timestampResolvers = {
  createdAt: (resource: RawTimestamped) => resource.created_at.toISOString(),
  updatedAt: (resource: RawTimestamped) => resource.updated_at.toISOString(),
};
