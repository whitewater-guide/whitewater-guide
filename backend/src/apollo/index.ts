// This should be imported separately, otherwise it creates circular dependencies
// export * from './loadGraphqlFile';
export * from './types';
export * from './enhancedResolvers';
export * from './listResolvers';
export * from './schema';
export * from './router';
export * from './formatError';
