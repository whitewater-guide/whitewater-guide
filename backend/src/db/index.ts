import db from './db';

export default db;
export * from './updatedAtTrigger';
export * from './types';
export * from './db';
