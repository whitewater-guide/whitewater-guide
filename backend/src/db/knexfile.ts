import * as knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: 'db',
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

const test: knex.Config = {
  client: 'pg',
  connection: {
    host: '127.0.0.1', // <------ Tests are run on host machine, not inside docker!
    port: 5432,
    database: 'test',
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/test',
  },
  debug: false, // process.env.DATABASE_DEBUG === 'true',
} as any;

const configs: { [env: string]: knex.Config } = {
  development,
  test,
};

export default configs;
