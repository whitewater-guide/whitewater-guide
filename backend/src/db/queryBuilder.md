# Query builder

Query builder is used to build deeply nested sql queries to match graphql queries

# How it works

![Schema](./docs/queryBuilder.jpg)

Consider following example graphql query:
```$graphql
query {
    sources {
        id
        name
        regions { // <--- connection to regions
            nodes {
                id
                name
            }
            count
        }
    }
}
```

1. Start building SQL query with some table. This is `root query`
2. Walk query tree and determine root field names. Some of them are simple columns, 
they are added to root query's `SELECT` clause. See `getPrimitives`
3. Others are connection fields. For each connection field `attachConnection` function runs.
As a result:
    1. CTE to select connection nodes is added to root query. This CTE is a `root query itself`.
    `build` property of connection map is used to build this CTE. *(2 on the picture)*
    2. Connection field is added to `SELECT` clause of `root query` as a subquery
    alongside primitive columns.
    This subquery is build with `buildConnectionColumn` function. *(3 on the picture)*
        1. In turn, this connection column subquery is composed of two parts
        2. The first part is another, inner CTE, to select and count nodes from outer CTE.
        Using `join` property of connection map *(4 on the picture)*, inner CTE is filtered to get connected to 
        table from `root query`. This is done by simply adding `WHERE` clause for one-to-many connections
        or by joining junction table and adding `WHERE` clause for many-to-many connections.
        It can also paginate (via `LIMIT/OFFSET`).
        3. The connection column is produced by building json from inner CTE.
        This is done in `buildConnectionJSONQuery` function. *(5 on the picture)*
4. (Not on the picture) Some others are one-to-one relations, sometimes indirect, like `section`->`region`. 
For them we want to build queries like
    ```$sql
    WITH rivers_ref as (
      SELECT * from rivers_view
      WHERE rivers_view.language = 'en'
    )
    SELECT
      sections_view.*,
      (SELECT to_json(rivers_ref.*) from rivers_ref where rivers_ref.id = sections_view.river_id) as river
    from sections_view
    WHERE sections_view.language = 'en'
    ```
    
    This is done in `attachOneToOnes` function 
5. As a result, a massive nested sql query is produced *(1 on the picture)*