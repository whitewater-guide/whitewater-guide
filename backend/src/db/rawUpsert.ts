import Knex = require('knex');

export async function rawUpsert<T>(db: Knex, query: string, queryName?: string): Promise<T> {
  const key: string = queryName || (query.split('SELECT ')[1]).split('(')[0];
  const result = await db.raw(query);
  return result.rows[0][key][0];
}
