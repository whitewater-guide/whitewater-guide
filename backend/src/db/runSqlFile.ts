import * as fs from 'fs';
import Knex = require('knex');

export async function runSqlFile(db: Knex, path: string) {
  const upsertRegion = await fs.readFileSync(path).toString();
  await db.schema.raw(upsertRegion);
}
