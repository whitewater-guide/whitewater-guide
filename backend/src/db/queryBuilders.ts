import { GraphQLResolveInfo } from 'graphql';
import * as Knex from 'knex';
import { snakeCase } from 'lodash';
import { Context } from '../apollo';
import graphqlFields = require('graphql-fields');
import db from './db';

type ColumnMap<T> = {
  [field in keyof T]?: (context?: Context) => any;
};

type ConnectionsMap<T> = {
  [field in keyof T]?: { build: Builder, join: Joiner } | null;
};

type Builder = (options: Partial<QueryBuilderOptions<any>>) => Knex.QueryBuilder;
type Joiner = (table: string, query: Knex.QueryBuilder) => Knex.QueryBuilder;

export const getPrimitives = <T>(
  topLevelFields: Array<keyof T>,
  prefix: string,
  context?: Context,
  connections?: Array<keyof T>,
  customMap?: ColumnMap<T>,
): string[] => {
  return topLevelFields.reduce((result, field) => {
    // connection types and __typename are ignored
    if (field === '__typename' || connections && connections.includes(field)) {
      return result;
    }
    // Custom map allows to conditionally drop some statements based on context
    if (customMap && field in customMap) {
      const mapped = customMap[field]!(context);
      return mapped ?  [...result, `${prefix}.${mapped}`] : result;
    }
    return [...result, `${prefix}.${snakeCase(field)}`];
  }, []);
};

export interface QueryBuilderOptions<T> {
  info?: GraphQLResolveInfo;
  fieldsTree?: {[key: string]: any};
  context: Context,
  table: string;
  tableAlias?: string;
  customFieldMap?: ColumnMap<T>;
  connections?: ConnectionsMap<T>;
  knex?: Knex;
  language?: string;
  orderBy?: string;
  id?: string;
}

export const buildRootQuery = <T>(options: QueryBuilderOptions<T>): Knex.QueryBuilder => {
  const {
    info,
    fieldsTree,
    context,
    table,
    tableAlias,
    customFieldMap = {},
    connections = {},
    knex = db(),
    language = 'en',
    orderBy = 'name',
    id,
  } = options;
  const alias = tableAlias || table;
  let result = knex.from(tableAlias ? `${table} AS ${tableAlias}` : table);
  if (!fieldsTree && !info) {
    throw new Error('Provide either resolver info or parsed fields tree from resolver info');
  }
  const rootFieldsTree = fieldsTree || graphqlFields(info!);
  const topLevelFields: Array<keyof T> = Object.keys(rootFieldsTree) as any;
  const primitiveColumns = getPrimitives<T>(
    topLevelFields,
    alias,
    context,
    Object.keys(connections) as any,
    customFieldMap,
  );
  result = result.select(primitiveColumns);
  Object.entries(connections).forEach(([connectionName, value]) => {
    if (value && rootFieldsTree[connectionName]) {
      const { build, join } = value;
      attachConnection({
        knex,
        query: result,
        build,
        join,
        context,
        language,
        name: connectionName,
        fieldsTree: rootFieldsTree[connectionName],
      });
    }
  });
  if (id) {
    result = result.where(`${alias}.id`, id);
  }
  if (language) {
    result = result.where(`${alias}.language`, language);
  }
  if (orderBy) {
    result = result.orderBy(`${alias}.${orderBy}`);
  }
  return result;
};

/**
 * This will build a knex query which selects json object with fields 'nodes' (optional) and 'count' required
 * 'nodes' will contain json array of all rows (this will also include redundant count in each item)
 * @param {string} table - CTE which has 'count' column made with window function
 * @param {boolean} includeNodes - include nodes or just count them?
 * @param {Knex.Transaction | Knex} knex
 * @returns {Knex.QueryBuilder}
 */
export const buildConnectionSubquery = (table: string, includeNodes = true, knex = db()) => {
  // Assumption: we can request count without nodes, but if we request nodes we also must request count
  const nodes = includeNodes ? knex.select(knex.raw(`json_agg(${table}.*)`)).from(table) : null;
  const count = includeNodes ?
    knex.select(`${table}.count`).from(table).limit(1) : // count comes from window function column
    knex.count(`${table}.*`).from(table);
  const parts = [`'count'`, `(${count.toQuery()})`];
  if (nodes) {
    parts.push(`'nodes'`, `(${nodes.toQuery()})`);
  }
  return knex.select(knex.raw(`json_build_object(${parts.join(', ')})`));
};

export interface ConnectionBuilderOptions {
  table: string;
  join: Joiner;
  includeNodes?: boolean;
  knex?: Knex;
  limit?: number;
  offset?: number;
}

export const buildConnectionField = (options: ConnectionBuilderOptions) => {
  const {
    table,
    join,
    includeNodes = true,
    knex = db(),
    limit,
    offset,
  } = options;
  const result = buildConnectionSubquery(`${table}_internal`, includeNodes, knex);
  return result.with(`${table}_internal`, (db2) => {
    let cte = db2.select(`${table}.*`, knex.raw(`count(${table}.*) OVER()`)).from(table);
    cte = join(table, cte);
    if (limit) {
      cte = cte.limit(limit);
      if (offset) {
        cte = cte.offset(offset);
      }
    }
    return cte;
  });
};

interface AttachConnectionOptions {
  query: Knex.QueryBuilder;
  context: Context;
  name: string;
  build: Builder;
  join: Joiner;
  language: string;
  fieldsTree: {[key: string]: any};
  knex?: Knex;
}

export const attachConnection = (options: AttachConnectionOptions) => {
  const {
    query,
    language,
    context,
    name,
    build,
    join,
    fieldsTree,
    knex,
  } = options;
  const connection = buildConnectionField({
    knex,
    table: `${name}_connection`,
    join,
    includeNodes: !!fieldsTree.nodes,
  });
  return query.with(`${name}_connection`, knex!.raw(build({
      fieldsTree: fieldsTree.nodes || { id: {} },
      context,
      language,
      knex,
    }).toQuery()),
  ).select(knex!.raw(`(${connection.toQuery()}) as ${name}`));
};
