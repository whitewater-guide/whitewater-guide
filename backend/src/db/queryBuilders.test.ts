import { Context } from '../apollo';
import { adminContext, anonContext } from '../test/context';
import { isAdmin } from '../ww-commons';
import * as Knex from 'knex';
import db from './db';
import {
  attachConnection,
  buildConnectionColumn,
  buildConnectionJSONQuery,
  buildRootQuery,
  ConnectionBuilderOptions,
  getPrimitives,
  QueryBuilderOptions,
} from './queryBuilders';
import gqf = require('graphql-fields');

jest.mock('graphql-fields', () => jest.fn());
const graphqlFields: jest.Mock<any> = gqf as any;

describe('getPrimitives', () => {
  const topLevelFields = ['__typename', 'camelCase', 'snake_case', 'regular', 'everyone', 'admin', 'connection'];
  const customMap = {
    admin: (c: Context) => isAdmin(c.user) ? 'admin' : null,
  };
  const prefix = 'tablename';

  it('should prefix fields', () => {
    const result = getPrimitives<any>(topLevelFields, prefix);
    result.forEach(field => expect(field.indexOf(prefix)).toBe(0));
  });

  it('should omit __typename', () => {
    const result = getPrimitives<any>(topLevelFields, prefix);
    expect(result).not.toContain('tablename.__typename');
  });

  it('should convert graphql camelCase to postgres snake_case', () => {
    expect(getPrimitives<any>(topLevelFields, prefix)).toContain('tablename.camel_case');
  });

  it('should not include connections', () => {
    expect(getPrimitives<any>(topLevelFields, prefix, undefined, ['connection']))
      .not.toContain('tablename.connection');
  });

  it('should omit based on context', () => {
    const userResult = getPrimitives<any>(topLevelFields, prefix, anonContext, ['connection'], customMap);
    expect(userResult).not.toContain('tablename.admin');
    const adminResult = getPrimitives<any>(topLevelFields, prefix, adminContext, ['connection'], customMap);
    expect(adminResult).toContain('tablename.admin');
  });

  it('should match snapshot', () => {
    const userResult = getPrimitives<any>(topLevelFields, prefix, anonContext, ['connection'], customMap);
    expect(userResult).toMatchSnapshot();
  });

});

describe('buildRootQuery', () => {
  const commonOptions: QueryBuilderOptions<any> = {
    info: {} as any,
    context: anonContext,
    knex: db(true),
    table: 'tbl',
    language: '', // empty string to avoid default language
    orderBy: '', // empty string to avoid sort
  };

  it('should construct basic query', () => {
    graphqlFields.mockReturnValueOnce({ bC: {} });
    const query = buildRootQuery(commonOptions);
    expect(query.toQuery()).toEqual('select "tbl"."b_c" from "tbl"');
  });

  it('should use fields tree if provided', () => {
    const query = buildRootQuery({ ...commonOptions, info: undefined, fieldsTree: { bC: {} } });
    expect(query.toQuery()).toEqual('select "tbl"."b_c" from "tbl"');
  });

  it('should throw if no info or fieldsTree provided', () => {
    expect(() => buildRootQuery({ ...commonOptions, info: undefined })).toThrow();
  });

  it('should use english by default', () => {
    graphqlFields.mockReturnValueOnce({ name: {} });
    const query = buildRootQuery({ ...commonOptions, language: undefined });
    expect(query.toQuery()).toEqual('select "tbl"."name" from "tbl" where "tbl"."language" = \'en\'');
  });

  it('should order by name by default', () => {
    graphqlFields.mockReturnValueOnce({ name: {} });
    const query = buildRootQuery({ ...commonOptions, orderBy: undefined });
    expect(query.toQuery()).toEqual('select "tbl"."name" from "tbl" order by "tbl"."name" asc');
  });

  it('should specify sort order', () => {
    graphqlFields.mockReturnValueOnce({ a: {}, b: {} });
    const query = buildRootQuery({ ...commonOptions, orderBy: 'b' });
    expect(query.toQuery()).toEqual('select "tbl"."a", "tbl"."b" from "tbl" order by "tbl"."b" asc');
  });

  it('should alias table', () => {
    graphqlFields.mockReturnValueOnce({ b: {} });
    const query = buildRootQuery({ ...commonOptions, tableAlias: 'als' });
    expect(query.toQuery()).toEqual('select "als"."b" from "tbl" as "als"');
  });

  it('should filter by id and language', () => {
    graphqlFields.mockReturnValueOnce({ b: {}, id: {}, language: {} });
    const query = buildRootQuery({ ...commonOptions, id: 'foo', language: 'en' });
    // tslint:disable-next-line:max-line-length
    const qstring = 'select "tbl"."b", "tbl"."id", "tbl"."language" from "tbl" where "tbl"."id" = \'foo\' and "tbl"."language" = \'en\'';
    expect(query.toQuery()).toEqual(qstring);
  });

  it('should respect custom fields map', () => {
    graphqlFields.mockReturnValueOnce({ b: {}, camelCase: {}, c: {} });
    const customFieldMap = {
      camelCase: () => 'snake_case',
      c: (c: Context) => isAdmin(c.user) ? 'c' : null,
    };
    const query = buildRootQuery({ ...commonOptions, customFieldMap });
    expect(query.toQuery()).toEqual('select "tbl"."b", "tbl"."snake_case" from "tbl"');
  });

  it('should exclude connections', () => {
    graphqlFields.mockReturnValueOnce({ b: {}, connection: {}, c: {} });
    const connections = { connection: null };
    const query = buildRootQuery({ ...commonOptions, connections });
    expect(query.toQuery()).toEqual('select "tbl"."b", "tbl"."c" from "tbl"');
  });

  it('should attach connections', () => {
    graphqlFields.mockReturnValueOnce({ b: {}, c: { id: {}, fk: {} } });
    const connections = { c: {
      build: () => db(true).select('*').from('k'),
      join: (t: any, q: any) => q.where(`${t}.fk`, '=', db(true).raw('??', ['tbl.id'])),
    }};
    const query = buildRootQuery({ ...commonOptions, connections });
    expect(query.toQuery()).toMatchSnapshot();
  });
});

describe('buildConnectionJSONQuery', () => {
  it('should build correct sql without nodes', () => {
    const result = buildConnectionJSONQuery('regions_connection_internal', false, db(true));
    expect(result.toQuery()).toMatchSnapshot();
  });

  it('should build correct sql with nodes', () => {
    const result = buildConnectionJSONQuery('regions_connection_internal', true, db(true));
    expect(result.toQuery()).toMatchSnapshot();
  });
});

describe('buildConnectionColumn', () => {
  const options: ConnectionBuilderOptions = {
    knex: db(true),
    table: 'tbl',
    join: (table, query) => {
      return query.where(`${table}.fk`, '=', db(true).raw('??', ['src.id']));
    },
  };

  it('should build simple connection field select', () => {
    const result = buildConnectionColumn(options).toQuery();
    expect(result).toMatchSnapshot();
  });

  it('should include limit and offset', () => {
    const result = buildConnectionColumn({ ...options, limit: 10, offset: 20 }).toQuery();
    expect(result).toMatchSnapshot();
  });

});

describe('attachConnection', () => {
  const rootQuery = db(true).select(['source.id', 'sources.name']).from('sources');
  const builder = (opts: QueryBuilderOptions<any>) => buildRootQuery({
    ...opts,
    table: 'regions',
    language: '', // shorten query
    orderBy: '', // shorten query
  });
  // one-to-many relation, for shorter query
  const joiner = (table: string, query: Knex.QueryBuilder) =>
    query.where(`${table}.source_id`, '=', db(true).raw('??', ['source.id']));

  let options: any;
  beforeEach(() => {
    options = {
      query: rootQuery.clone(),
      language: '', // for shorter query
      fieldsTree: { nodes: { id: {}, name: {} }, count: {} },
      context: adminContext,
      name: 'regions',
      knex: db(true),
      join: joiner,
      build: builder,
    };
  });

  it('should modify root query', () => {
    const result = attachConnection(options);
    expect(result.toQuery()).toMatchSnapshot();
  });

  it('should query ids if nodes are not required', () => {
    const result = attachConnection({ ...options, fieldsTree: { count: {} } });
    expect(result.toQuery()).toMatchSnapshot();
  });
});
