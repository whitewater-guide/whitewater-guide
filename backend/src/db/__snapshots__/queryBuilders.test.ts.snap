// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachConnection should modify root query for many-to-may connections 1`] = `
with regions_connection as (
    select
        regions.id,
        regions.name
    from
        regions
    where
        regions.language = 'en'
    order by
        regions.name asc
)
select
    source.id,
    sources.name, (with regions_connection_internal as (
            select
                regions_connection.*,
                COALESCE(count(regions_connection.*)
                    OVER (),
                    0)
            from
                regions_connection
            inner join sources_regions on regions_connection.id = sources_regions.region_id
            where
                sources_regions.source_id = source.id
)
        select
            json_build_object('count', (
                    select
                        regions_connection_internal.count from regions_connection_internal
                    limit 1),
                'nodes', (
                    select
                        COALESCE(json_agg (regions_connection_internal.*),
                            '[]'::json)
                    from
                        regions_connection_internal))) as regions
from
    sources

`;

exports[`attachConnection should modify root query for one-to-many connections 1`] = `
with regions_connection as (
    select
        regions.id,
        regions.name,
        regions.source_id
    from
        regions
    where
        regions.language = 'en'
    order by
        regions.name asc
)
select
    source.id,
    sources.name, (with regions_connection_internal as (
            select
                regions_connection.*,
                COALESCE(count(regions_connection.*)
                    OVER (),
                    0)
            from
                regions_connection
            where
                regions_connection.source_id = source.id
)
        select
            json_build_object('count', (
                    select
                        regions_connection_internal.count from regions_connection_internal
                    limit 1),
                'nodes', (
                    select
                        COALESCE(json_agg (regions_connection_internal.*),
                            '[]'::json)
                    from
                        regions_connection_internal))) as regions
from
    sources

`;

exports[`attachConnection should query ids if nodes are not required 1`] = `
with regions_connection as (
    select
        regions.id
    from
        regions
    where
        regions.language = 'en'
    order by
        regions.name asc
)
select
    source.id,
    sources.name, (with regions_connection_internal as (
            select
                regions_connection.*,
                COALESCE(count(regions_connection.*)
                    OVER (),
                    0)
            from
                regions_connection
)
        select
            json_build_object('count', (
                    select
                        regions_connection_internal.count from regions_connection_internal
                    limit 1))) as regions
from
    sources

`;

exports[`build list query should apply offset and limit 1`] = `
select
    regions_view.id,
    regions_view.name,
    count(regions_view.*)
    OVER ()
from
    regions_view
where
    regions_view.language = 'en'
order by
    regions_view.name asc
limit 10 offset 20

`;

exports[`build list query should query count only 1`] = `
select
    count(*)
from
    regions_view
where
    regions_view.language = 'en'

`;

exports[`buildConnectionColumn should build simple connection field select 1`] = `
with tbl_internal as (
    select
        tbl.*,
        COALESCE(count(tbl.*)
            OVER (),
            0)
    from
        tbl
    where
        tbl.fk = src.id
)
select
    json_build_object('count', (
            select
                tbl_internal.count from tbl_internal
            limit 1),
        'nodes', (
            select
                COALESCE(json_agg (tbl_internal.*),
                    '[]'::json)
            from
                tbl_internal))

`;

exports[`buildConnectionColumn should include limit and offset 1`] = `
with tbl_internal as (
    select
        tbl.*,
        COALESCE(count(tbl.*)
            OVER (),
            0)
    from
        tbl
    where
        tbl.fk = src.id
    limit 10 offset 20
)
select
    json_build_object('count', (
            select
                tbl_internal.count from tbl_internal
            limit 1),
        'nodes', (
            select
                COALESCE(json_agg (tbl_internal.*),
                    '[]'::json)
            from
                tbl_internal))

`;

exports[`buildConnectionJSONQuery should build correct sql with nodes 1`] = `
select
    json_build_object('count', (
            select
                regions_connection_internal.count from regions_connection_internal
            limit 1),
        'nodes', (
            select
                COALESCE(json_agg (regions_connection_internal.*),
                    '[]'::json)
            from
                regions_connection_internal))

`;

exports[`buildConnectionJSONQuery should build correct sql without nodes 1`] = `
select
    json_build_object('count', (
            select
                regions_connection_internal.count from regions_connection_internal
            limit 1))

`;

exports[`buildRootQuery should alias table 1`] = `
select
    als.b
from
    tbl as als
where
    als.language = 'en'
order by
    als.name asc

`;

exports[`buildRootQuery should attach connections 1`] = `
with c_connection as (
    select
        *
    from
        k
)
select
    tbl.b, (with c_connection_internal as (
            select
                c_connection.*,
                COALESCE(count(c_connection.*)
                    OVER (),
                    0)
            from
                c_connection
            where
                c_connection.fk = tbl.id
)
        select
            json_build_object('count', (
                    select
                        c_connection_internal.count from c_connection_internal
                    limit 1))) as c
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should construct basic query 1`] = `
select
    tbl.b_c
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should exclude connections 1`] = `
select
    tbl.b,
    tbl.c
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should not include connection when not asked by graphql 1`] = `
select
    tbl.b,
    tbl.c
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should respect custom fields map 1`] = `
select
    tbl.b,
    tbl.snake_case
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should specify language, id, sort order 1`] = `
select
    tbl.b,
    tbl.id,
    tbl.language
from
    tbl
where
    tbl.id = 'foo'
    and tbl.language = 'en'
order by
    tbl.b asc

`;

exports[`buildRootQuery should use english by default 1`] = `
select
    tbl.name
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`buildRootQuery should use fields tree if provided 1`] = `
select
    tbl.b_c
from
    tbl
where
    tbl.language = 'en'
order by
    tbl.name asc

`;

exports[`getPrimitives should match snapshot 1`] = `
Array [
  "tablename.camel_case",
  "tablename.snake_case",
  "tablename.regular",
  "tablename.everyone",
]
`;
